### PR TITLE
Public API approval test

### PR DESCRIPTION
**Pull request by SeanFeldman**
Connects to #43

Added public API approval test to control what's exposed by the transport.

### CODE REVIEW

**Comment by SeanFeldman** [comment]
@Particular/azure-maintainers please review

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 1]
```diff
@@ -0,0 +1,215 @@
+﻿namespace NServiceBus.Azure.Transports.WindowsAzureStorageQueues
+{
+    
+    public class AzureMessageQueueCreator : NServiceBus.Transports.ICreateQueues
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 2]
```diff
@@ -0,0 +1,215 @@
+﻿namespace NServiceBus.Azure.Transports.WindowsAzureStorageQueues
+{
+    
+    public class AzureMessageQueueCreator : NServiceBus.Transports.ICreateQueues
+    {
+        public AzureMessageQueueCreator(Microsoft.WindowsAzure.Storage.Queue.CloudQueueClient client, NServiceBus.Azure.Transports.WindowsAzureStorageQueues.QueueAddressGenerator addressGenerator) { }
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 3]
```diff
@@ -0,0 +1,215 @@
+﻿namespace NServiceBus.Azure.Transports.WindowsAzureStorageQueues
+{
+    
+    public class AzureMessageQueueCreator : NServiceBus.Transports.ICreateQueues
+    {
+        public AzureMessageQueueCreator(Microsoft.WindowsAzure.Storage.Queue.CloudQueueClient client, NServiceBus.Azure.Transports.WindowsAzureStorageQueues.QueueAddressGenerator addressGenerator) { }
+        public async System.Threading.Tasks.Task CreateQueueIfNecessary(NServiceBus.Transports.QueueBindings queueBindings, string identity) { }
+    }
+    public class AzureStorageQueueInfrastructure : NServiceBus.Transports.TransportInfrastructure
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal (working on a pull to make the msmq counterpart internal)

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+        public override int GetHashCode() { }
+        public override string ToString() { }
+    }
+    public class CreateQueueClients : NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ICreateQueueClients
+    {
+        public CreateQueueClients() { }
+        public Microsoft.WindowsAzure.Storage.Queue.CloudQueueClient Create(NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ConnectionString connectionString) { }
+        public Microsoft.WindowsAzure.Storage.Queue.CloudQueueClient CreateRecevier(NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ConnectionString connectionString) { }
+    }
+    public class static DictionaryExtensions
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+    public sealed class ConnectionString : System.IEquatable<NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ConnectionString>
+    {
+        public readonly string Value;
+        public ConnectionString(string value) { }
+        public bool Equals(NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ConnectionString other) { }
+        public override bool Equals(object obj) { }
+        public override int GetHashCode() { }
+        public override string ToString() { }
+    }
+    public class CreateQueueClients : NServiceBus.Azure.Transports.WindowsAzureStorageQueues.ICreateQueueClients
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal and just keep the interface public

**Comment by SeanFeldman** [comment]
Those types should be definitely internalized.

@andreasohlund would it make sense to do the review you started on a separate PR? 
This PR was solely focusing on introducing APIApprover into ASQ.

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt (hunk 2 above)
Comment (SeanFeldman) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: I presume you refer to `QueueAddressGenerator` @andreasohlund

**Comment by SeanFeldman** [comment]
All methined above by @andreasohlund are internalized now. I have created a linked version of `QueueAddressGenerator` for ATT project since it's [used there](https://github.com/Particular/NServiceBus.AzureStorageQueues/blob/05b1b354822a76ae70fd096b8a56660ff5054b17/src/AcceptanceTests/ConfigureAzureStorageQueueTransport.cs#L68).

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+    }
+    public class MessageWrapperSerializer
+    {
+        public static readonly System.Lazy<NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapperSerializer> Json;
+        public static readonly System.Lazy<NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapperSerializer> Xml;
+        public MessageWrapperSerializer(System.Action<NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapper, System.IO.Stream> serialize, System.Func<System.IO.Stream, NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapper> deserialize) { }
+        public NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapper Deserialize(System.IO.Stream stream) { }
+        public void Serialize(NServiceBus.Azure.Transports.WindowsAzureStorageQueues.MessageWrapper wrapper, System.IO.Stream stream) { }
+    }
+    public class QueueAddressGenerator
```
Comment (andreasohlund) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Should be able to make this internal

**Comment by andreasohlund** [comment]
@Particular/transport-maintainers @Particular/persistence-maintainers what do you think of doing a similar exercise on your repos? (add ApiApprover and go through all public types an make as much as possible internal)

**Comment by DavidBoike** [comment]
Absolutely. Will raise issues for Raven/NH.

**Comment by tmasternak** [comment]
@andreasohlund :+1: - I'll create issue for SqlServer

**Comment by SeanFeldman** [comment]
If no objection (@andreasohlund), I'll :shipit:

File: src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt [hunk 7]
```diff
@@ -0,0 +1,153 @@
+﻿namespace NServiceBus.Azure.Transports.WindowsAzureStorageQueues.Config
```
Comment (melkio) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: Is it right this namespace? On ASB we use `NServiceBus.AzureServiceBus`. Could be `NServiceBus.AzureStorageQueue` more "consistent"?

Comment (SeanFeldman) [inline src/Tests/API/APIApprovals.ApproveAzureStorageQueueTransport.approved.txt]: It's a valid point @melkio. 

There are types that should be internalized (like `WellKnownConfigurationKeys`). For the Config namespace, I'm not sure what the original intent was. If it was not to pollute the higher level namespace with things like `ConnectionString` or `EnvelopeDeserializationFailed` exception (which also could be named to `EnvelopeDeserializationFailedException`), then I can see this as a legitimate thing.

Saying that, in my comment https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/56#issuecomment-200554967 I've asked if we should have a **separate PR**. Concern of this PR was to introduce Public API approval into the project.

**Comment by SeanFeldman** [comment]
@here, there's something about `When_timeout_dispatch_fails.Should_retry_and_move_to_error` that is flaky or the cleanup is not working properly. All tests/ATTs are passing locally. Will re-run. Perhaps third time's the charm.

**Comment by SeanFeldman** [comment]
There's [another PR](https://github.com/Particular/NServiceBus.AzureStorageQueues/pull/57) that is failing on the same test. I cannot reproduce this locally as all tests are green. I'd like to merge it. Any volunteers to push the button? If no responses are coming, I'll do it :)

**Comment by dvdstelt** [comment]
Most likely related to non empty queues. It didn't work properly and we changed it, but still isn't working correctly.

**Comment by Scooletz** [comment]
@dvdstelt @SeanFeldman If the error has no repro, but still, is reoccurring, imho we should create an issue for that. Could you create it?

**Comment by SeanFeldman** [comment]
Issue https://github.com/Particular/NServiceBus.AzureStorageQueues/issues/60 raised.

@Particular/azure-maintainers who'd like to press the 
image 
button?

**Comment by SeanFeldman** [comment]
Ping @Particular/azure-maintainers

**Comment by SeanFeldman** [comment]
Rebased

**Pull request merged**